src/MongoDB.Driver.Core/Core/Misc/StreamExtensionMethods.cs

```diff
@@ -41,13 +41,22 @@ public static async Task<int> ReadAsync(this Stream stream, byte[] buffer, int o
             var state = 1; // 1 == reading, 2 == done reading, 3 == timedout, 4 == cancelled
 
             var bytesRead = 0;
-            using (new Timer(_ => ChangeState(3), null, timeout, Timeout.InfiniteTimeSpan))
             using (cancellationToken.Register(() => ChangeState(4)))
             {
                 try
                 {
-                    bytesRead = await stream.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
-                    ChangeState(2); // note: might not actually go to state 2 if already in state 3 or 4
+                    var readTask = stream.ReadAsync(buffer, offset, count, cancellationToken);
+                    var timeoutTask = Task.Delay(timeout, cancellationToken);
+                    var completedTask = await Task.WhenAny(readTask, timeoutTask).ConfigureAwait(false);
+                    if (completedTask == readTask)
+                    {
+                        bytesRead = readTask.Result;
```

Comment:
If I'm not mistaken, this should be `GetAwaiter.GetResult()`, as far as I can see, otherwise we will catch `AggregationException` in case if ReadAsync throw any exception

Reply:
Correct. Will change in next commit if we decide to proceed.